lib/knapsack_pro/runners/queue/rspec_runner.rb

```diff
@@ -3,224 +3,287 @@
+          @function_core = function_core
+          @cli_args = prepare_cli_args(args)
+          @stream_error = stream_error
+          @stream_out = stream_out
+          @node_assigned_test_file_paths = []
+          @rspec_runner = nil # RSpec::Core::Runner is lazy initialized
+        end
 
-          if [
-              ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::RSpecQueueSummaryFormatter.to_s),
-              ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::TimeTracker.to_s),
-          ].all?
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L85
+        #
+        # @return [Fixnum] exit status code.
+        #   0 if all specs passed,
+        #   or the configured failure exit code (1 by default) if specs failed.
+        def run
+          pre_run_setup
+
+          if @rspec_runner.knapsack__wants_to_quit?
+            exit_code = @rspec_runner.knapsack__exit_early
+            Kernel.exit(exit_code)
             return
           end
 
-          unless ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::RSpecQueueSummaryFormatter.to_s)
-            ENV['SPEC_OPTS'] = "#{ENV['SPEC_OPTS']} --format #{KnapsackPro::Formatters::RSpecQueueSummaryFormatter}"
+          begin
+            exit_code = @rspec_runner.knapsack__run_specs(self)
+          rescue KnapsackPro::Runners::Queue::BaseRunner::TerminationError
+            @function_core.set_error_exit_code(@rspec_runner.knapsack__error_exit_code)
+            raise
+          rescue Exception => exception
+            KnapsackPro.logger.error("An unexpected exception happened. RSpec cannot handle it. The exception: #{exception.inspect}")
+            @function_core.log_exit_summary(@node_assigned_test_file_paths)
+            @function_core.set_error_exit_code(@rspec_runner.knapsack__error_exit_code)
+            raise
           end
 
-          unless ENV['SPEC_OPTS'].include?(KnapsackPro::Formatters::TimeTracker.to_s)
-            ENV['SPEC_OPTS'] = "#{ENV['SPEC_OPTS']} --format #{KnapsackPro::Formatters::TimeTracker}"
+          post_run_tasks(exit_code)
+        end
+
+        def with_batched_tests_from_queue
+          can_initialize_queue = true
+
+          loop do
+            handle_signal!
+            test_file_paths = pull_tests_from_queue(can_initialize_queue: can_initialize_queue)
+            can_initialize_queue = false
+
+            break if test_file_paths.empty?
+
+            with_queue_hooks(test_file_paths) do |wrapped_test_file_paths|
+              yield wrapped_test_file_paths
+            end
           end
         end
 
+        def handle_signal!
+          self.class.handle_signal!
+        end
+
+        def log_fail_fast_limit_met
+          @function_core.log_fail_fast_limit_met
+        end
+
         private
 
-        def self.adapter_class
-          KnapsackPro::Adapters::RSpecAdapter
+        def prepare_cli_args(args)
```

Comment:
It could. I did not want to test such a method, so I assumed it was good enough to keep it inside of a private method here. It's up to the queue_runner to decide what to use and in what order from the functional_core.

Reply:
Even if you don't want to test it, I don't see why we need this indirection.

Also, these 3 methods are only called here together. So it would make more sense for them to live in an abstraction (functional core) behind a single public method and have those 3 as private methods.

Also, the test would cover the single public method if feasible.

Reply:
done: https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237/commits/98196470af32b8762a1e6bb2aa1d11610bc762f1